Introduce tablets

tgrabiec opened the pull request:
This PR introduces an experimental feature called "tablets". Tablets are a way to distribute data in the cluster, which is an alternative to the current vnode-based replication. Vnode-based replication strategy tries to evenly distribute the global token space shared by all tables among nodes and shards. With tablets, the aim is to start from a different side. Divide resources of replica-shard into tablets, with a goal of having a fixed target tablet size, and then assign those tablets to serve fragments of tables (also called tablets). This will allow us to balance the load in a more flexible manner, by moving individual tablets around. Also, unlike with vnode ranges, tablet replicas live on a particular shard on a given node, which will allow us to bind raft groups to tablets. Those goals are not yet achieved with this PR, but it lays the ground for this. 

Things achieved in this PR:

  - You can start a cluster and create a keyspace whose tables will use tablet-based replication. This is done by setting `initial_tablets` option:
  
    ```
        CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy',
                        'replication_factor': 3, 
                        'initial_tablets': 8};
    ```

    All tables created in such a keyspace will be tablet-based.

    Tablet-based replication is a trait, not a separate replication strategy. Tablets don't change the spirit of replication strategy, it just alters the way in which data ownership is managed. In theory, we could use it for other strategies as well like EverywhereReplicationStrategy. Currently, only NetworkTopologyStrategy is augmented to support tablets.

  - You can create and drop tablet-based tables (no DDL language changes)
 
  - DML / DQL work with tablet-based tables

    Replicas for tablet-based tables are chosen from tablet metadata instead of token metadata 

Things which are not yet implemented:

  - handling of views, indexes, CDC created on tablet-based tables 
  - sharding is done using the old method, it ignores the shard allocated in tablet metadata 
  - node operations (topology changes, repair, rebuild) are not handling tablet-based tables 
  - not integrated with compaction groups
  - tablet allocator piggy-backs on tokens to choose replicas. Eventually we want to allocate based on current load, not statically

tgrabiec commented on query_ranges_to_vnodes.cc:
```diff
@@ -101,18 +101,6 @@ void query_ranges_to_vnodes_generator::process_one_range(size_t n, dht::partitio
-        if (is_empty) {
-            _i++;
-            return;
-        }
```
\cc @gusev-p

tgrabiec replied:
There is a rationale in the commit message: https://github.com/scylladb/scylladb/pull/13387/commits/64674e034234ef94ed01d8676b6602a35c5a8483

bhalevy replied:
doesn't this create a regression in the series that make it less bisectable?

avikivity replied:
I thought of the original patch as a fix, not an optimization, that just can't be triggered in the code as it is. So reverting it is dangerous.

tgrabiec replied:
> I thought of the original patch as a fix, not an optimization

I think it's just an optimization. @gusev-p Is it a fix?

gusev-p replied:
> I think it's just an optimization. @gusev-p Is it a fix?

The patch just prevents data queries with empty ranges. There is no reliance on the range being not empty though, so it's more of an optimisation, yes.

avikivity replied:
ok

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/404/

xemul commented on cql3/statements/select_statement.cc:
```diff
@@ -559,7 +560,9 @@ indexed_table_select_statement::do_execute_base_query(
     auto timeout = db::timeout_clock::now() + get_timeout(state.get_client_state(), options);
     uint32_t queried_ranges_count = partition_ranges.size();
-    query_ranges_to_vnodes_generator ranges_to_vnodes(qp.proxy().get_token_metadata_ptr(), _schema, std::move(partition_ranges));
+    auto&& table = qp.proxy().local_db().find_column_family(_schema);
```
nit: `qp.proxy().local_db()` can be shortened to `qp.db()`

tgrabiec replied:
`qp.db()` returns data_ditctionary::database, which holds just the schema.

xemul commented on db/system_keyspace.cc:
```diff
@@ -1015,6 +1018,28 @@ schema_ptr system_keyspace::broadcast_kv_store() {
+schema_ptr system_keyspace::tablets() {
+    static thread_local auto schema = [] {
+        // FIXME: Allow UDTs in system keyspace:
+        // CREATE TYPE tablet_replica (replica_id uuid, shard int);
```
The new system.table needs a description in docs/dev/system_keyspace.md

xemul commented on main.cc:
```diff
@@ -1148,6 +1148,11 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+
+            ss.invoke_on_all([&] (service::storage_service& ss) {
+                ss.set_query_processor(qp.local());
+            }).get();
```
This is needed to let `storage_service::load_tablet_metadata()` below call `replica::read_tablet_metadata()` which, in turn (https://github.com/scylladb/scylladb/pull/13387/commits/29fd907dee0b686a5aaef45e1718313838a7f267) needs to do `qp.query_internal("select * from system.tablets", ...)`.

This can solved by keeping at least the `qp.query_internal("select * from system.tablets", ...)` part in `system_keyspace::load_tablets()` non-static method. The storage_service already has system_keyspace instance reference, so it will be able to call the proposed load_tablets() w/o the need to plug query processor into itself.

xemul replied:
If not system keyspace, everything that needs to keep data in system.* tables depends on query_processor (or qctx) and that's not any better.

System keyspace, in turn, only depends on query processor and database, but the latter can be removed.

tgrabiec replied:
I'm not thinking about run-time dependencies but code dependencies. system_keyspace.hh includes now the raft stuff, topology changes stuff, etc. It would now also include tablets.hh. Anything which deals with system keyspace, and includes system_keyspace.hh, would transitively pull these too. It's becoming a central hub for all the features.

gleb-cloudius replied:
Code dependencies can be fixed by splitting system_keyspace.[hc] into multiple files, but preserving runtime dependency graph, no?

tgrabiec replied:
Isn't what you suggesting equivalent to putting component-specific code in the component-specific .cc/.hh, like in this patch? https://github.com/scylladb/scylladb/pull/13387/commits/9d786c1ebc303373bf739a5d47ec2eeb5669b3b8

xemul commented on main.cc:
```diff
@@ -1111,6 +1112,14 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
                 forward_service.stop().get();
             });
 
+            distributed<service::tablet_allocator> the_tablet_allocator;
```
nit: The `the_` prefix is in use by `debug` namespace to store global pointers to services for scylla-gdb.py and whoever tries to debug scylla needs. So this `the_` prefix in local variable name looks confusing

tgrabiec replied:
Ok, I'll change it.

raphaelsc commented on dht/token.hh:
```diff
@@ -218,8 +218,18 @@ inline bool operator<=(const token& t1, const token& t2) { return std::rel_ops::
 
+// Returns a successor for token t.
+// Wraps around if there is no successor.
+//
```
see:
```
class token {
    // INT64_MIN is not a legal token, but a special value used to represent
    // infinity in token intervals.
    // If a token with value INT64_MIN is generated by the hashing algorithm,
    // the result is coerced into INT64_MAX.
    // (So INT64_MAX is twice as likely as every other token.)
    static inline int64_t normalize(int64_t t) {
        return t == std::numeric_limits<int64_t>::min() ? std::numeric_limits<int64_t>::max() : t;
    }
```

so next_token() call on _data == INT64_MAX will result in INT64_MIN (which is considered an illegal token).

but we don't want to loop around INT64_MAX, if we normalize it like above.

so perhaps we want to just skip INT64_MIN, to continue the iteration from the bottom of the valid token range.

tgrabiec replied:
Yes. I'll remove the wrap around thing and make it UB instead.

bhalevy commented on locator/abstract_replication_strategy.hh:
```diff
@@ -128,11 +128,48 @@ public:
 };
 
+using replication_strategy_ptr = seastar::shared_ptr<const abstract_replication_strategy>;
+using mutable_replication_strategy_ptr = seastar::shared_ptr<abstract_replication_strategy>;
```
let's split the above into a separate patch.
it seems orthogonal to reintroducing an abstract effective_replication_map

tgrabiec replied:
Ok

bhalevy commented on locator/abstract_replication_strategy.hh:
```diff
@@ -128,11 +128,48 @@ public:
+/// \brief Represents effective replication (assignment of replicas to keys).
+///
+/// It's a result of application of a given replication strategy
+/// over a particular token metadata version, for a given table.
```
using a given set of replication strategy options.

bhalevy replied:
not to confuse with a "pure" replication strategy which is a synonym for the class/type.

bhalevy commented on locator/abstract_replication_strategy.hh:
```diff
@@ -128,11 +128,48 @@ public:
+    token_metadata_ptr _tmptr;
+    size_t _replication_factor;
+public:
+    effective_replication_map(replication_strategy_ptr, token_metadata_ptr, size_t replication_factor);
```
`noexcept` right? (token_effective_replication_map ctor is noexcept today)

bhalevy replied:
Not any more. storage_service::replicate_to_all_cores handles exceptions when creating / copying token_metadata and effective_replication_map.
We're in disagreement on this.  I think that noexcept is beneficial when possible, not only when required.

bhalevy replied:
We shouldn't block the PR on this, but since `effective_replication_map` was already marked noexcept I suggest not to change that "on the fly" without verifying that it doesn't expose any caller that relies on it.
That's the point behind noexcept.  It shouldn't be changed on a whim.

tgrabiec replied:
> It shouldn't be changed on a whim.

True, I'll preserve the noexcept.

bhalevy commented on locator/abstract_replication_strategy.hh:
```diff
@@ -128,11 +128,48 @@ public:
+    const token_metadata& get_token_metadata() const noexcept { return *_tmptr; }
+    const token_metadata_ptr& get_token_metadata_ptr() const noexcept { return _tmptr; }
+    const topology& get_topology() const noexcept { return _tmptr->get_topology(); }
+    size_t get_replication_factor() const { return _replication_factor; }
```
noexcept

bhalevy commented on locator/abstract_replication_strategy.cc:
let's switch the order then and make this change before reintroducing the abstract effective_replication_map

tgrabiec replied:
github is not showing me context for this comment. What change?

bhalevy replied:
treewide: Use replication_strategy_ptr as a shorter name for abstract…
…_replication_strategy::ptr_type

tgrabiec replied:
Why do you think we should change the order?

bhalevy replied:
I think it belong with https://github.com/scylladb/scylladb/pull/13387/files#r1155252465
But we can have a small preliminary change to introduce the shorthand types
and then keep the patches in the current order.  This isn't a deal breaker :)

tgrabiec replied:
I addressed that suggestion and replication_strategy_ptr is introduced in the rename patch. 

If that's fine I'd rather not change the order because it creates rebase headache.

bhalevy replied:
sure, no problem

bhalevy commented on replica/database.cc:
```diff
@@ -1144,7 +1144,8 @@ std::vector<sstring> database::get_non_local_strategy_keyspaces() const {
     for (auto const& i : _keyspaces) {
-        if (i.second.get_replication_strategy().get_type() != locator::replication_strategy_type::local) {
+        auto&& rs = i.second.get_replication_strategy();
+        if (rs.get_type() != locator::replication_strategy_type::local && !rs.is_per_table()) {
```
maybe the function should be renamed too, since `non_local_strategy` technically includes per-table strategies too that are non-local.
e.g. `get_non_local_token_based_strategy_keyspaces`

tgrabiec replied:
Agreed

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -115,8 +115,7 @@ class token_metadata_impl final {
 
-    boost::iterator_range<token_metadata::tokens_iterator> ring_range(
-        const std::optional<dht::partition_range::bound>& start) const;
+    boost::iterator_range<token_metadata::tokens_iterator> ring_range(dht::ring_position_view pos) const;
```
nit: s/pos/start/ as in implementation

bhalevy commented on locator/token_range_splitter.hh:
```diff
@@ -0,0 +1,42 @@
+    virtual std::optional<dht::token> next_token() = 0;
+};
+
+}
```
nit: please add a `// namespace locator` trailing comment for future generations
(and a newline)

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+    return &i->second;
+}
+
+}
```
nit: namespace comment + newline

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+seastar::logger tablet_logger("tablets");
+
+std::ostream& operator<<(std::ostream& out, const tablet_replica& r) {
+    return out << r.host << ":" << r.shard;
```
@tchaikov would ask for a fmt::formatter specialization as this seems to be the new convention.

@tchaikov I don't think it's documented anywhere yet.
How about adding something under docs/dev?
There should probably be also a higher-level doc pointing to it, though I'm not sure which.

tgrabiec replied:
Is this a blocker?

tchaikov replied:
@bhalevy @tgrabiec yeah, but we can always fix it latter. i don't think it's a blocker. regarding the documentation, will come up with a proposal in this week. it will likely based on the issue of #13245 .

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+// Different tablets in subsequent token metadata version can have the same tablet_id.
+// When splitting a tablet, one of the new tablets (in the new token metadata version)
+// will have the same tablet_id as the old one.
+using tablet_id = size_t;
```
nit: typically, we use `_id` for unique uuid-based identifiers.
I think `table_idx` would better describe what it stands for.

bhalevy replied:
Also, it should be a tagged integer as proposed in https://github.com/scylladb/scylladb/pull/12966/commits/abb832238c015ab7428d95d363f058a923c7e352

avikivity replied:
We also have shard_id.

avikivity replied:
There's `enum class : size_t`, no need for tagged_integral.

bhalevy replied:
> We also have shard_id.

I think we would eventually also want to identify tablet replicas globally by a uuid,
so I think the in-memory tablet_id would be confusing for that reason.

tgrabiec replied:
> I think we would eventually also want to identify tablet replicas globally by a uuid,
> so I think the in-memory tablet_id would be confusing for that reason.

We can introduce tablet_replica_id then. I think it's a matter of terminology, what a "tablet" is. It could be also confusing to call the tablet replica by "tablet_id" if you don't know what "tablet" refers to. I propose to stick to the terminology where "tablet" is the thing at the top of the hierarchy, unless the context says otherwise. If there is no context, we use "tablet replica" for tablet replica.

bhalevy replied:
okay :crossed_fingers:

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+
+/// Stores information about a single tablet.
+struct tablet_info {
+    tablet_replica_set replicas;
```
since the tablet must not be replicated on multiple shards on the same host, how about using a `std::unordered_map<host_id, shard_id>`?
Does the order of replicas matter?

bhalevy replied:
If using a vector I'd add some checks in non-release modes around uniqueness of the host_id on the insertion and/or retrieval paths.

tgrabiec replied:
Such checks probably belong to the tablet allocator which produces these, and involve not only uniqueness of the host, but also racks and DCs. There's `endpoints_check()` in network_topology_strategy_test.cc. Maybe each replication strategy should provide its sanity checks? 

I think this can be done incrementally.

avikivity replied:
agree

bhalevy commented on dht/token.cc:
```diff
@@ -288,4 +288,14 @@ compaction_group_of(unsigned most_significant_bits, const token& t) {
     __builtin_unreachable();
 }
 
+token last_token_of_compaction_group(unsigned most_significant_bits, size_t group) {
```
Should it be used in `dht::split_token_range_msb`?

bhalevy replied:
the latter is called by `table::make_compaction_groups`

tgrabiec replied:
Yes, split_token_range_msb() can be implemented with this.

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+const tablet_transition_info* tablet_map::get_tablet_transition_info(tablet_id id) const {
+    auto i = _transitions.find(id);
+    if (i == _transitions.end()) {
+        return nullptr;
```
why does `get_tablet_info` return a reference and throws if `tablet_id` isn't found
but here we return a possibly-null pointer?

tgrabiec replied:
Because transition info is optional, while tablet info is not.

tgrabiec replied:
get_tablet_info() throws if there is an out of range access. This could throw in that case as well, but it doesn't seem necessary, it fits with the semantics to treat this as a missing transition.

tgrabiec replied:
tablet_id has a scope of a tablet map, and tablet map contains info about all tablets in the scope (whole ring). So calling get_tablet_info() with an id which is not present does not mean that you're trying to access a tablet for which the map doesn't contain a mapping, it means the id is wrong and meaningless (it's always a logic error).

bhalevy replied:
this merits some good ol'e doc comments

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+    explicit tablet_map(size_t tablet_count);
+
+    tablet_id get_tablet_id(token) const;
+    const tablet_info& get_tablet_info(tablet_id) const;
```
see question above regarding returning a reference vs. a pointer.
If we end up keeping this behavior, let's document the exception thrown when tablet_id isn't found.

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+
+    tablet_id get_tablet_id(token) const;
+    const tablet_info& get_tablet_info(tablet_id) const;
+    const tablet_transition_info* get_tablet_transition_info(tablet_id) const;
```
here too, document when nullptr is returned.

tgrabiec replied:
done

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+    const tablet_transition_info* get_tablet_transition_info(tablet_id) const;
+    dht::token get_last_token(tablet_id id) const;
+    dht::token get_first_token(tablet_id id) const;
+    dht::token_range get_token_range(tablet_id id) const;
```
why do we need both get_first, get_last, and get_token_range?
isn't the latter sufficient for all use cases?

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+dht::token tablet_map::get_first_token(tablet_id id) const {
+    if (id == 0) {
+        // dht::token::normalize() does not allow std::numeric_limits<int64_t>::min()
+        return dht::token(dht::token_kind::key, std::numeric_limits<int64_t>::min() + 1);
```
the `std::numeric_limits<int64_t>::min() + 1` logic belong in the dht layer, not here.
it's independent of tablets.

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+
+dht::token_range tablet_map::get_token_range(tablet_id id) const {
+    if (id == 0) {
+        return dht::token_range::make({dht::minimum_token(), false}, {get_last_token(id), true});
```
why not simply `return dht::token_range::make({get_first_token(id), true}, {get_last_token(id), true})` in all cases?
then we can make `get_{first,last}_token` private and expose only `get_token_range()`.

tgrabiec replied:
This is horrible, I know.

bhalevy replied:
Hmm, looks like I broke that undocumented convention in dht::split_token_range_msb (called from table::make_compaction_groups to set the compaction groups ranges).
Maybe the code that assumes the above is not executed in any of our tests.

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+            return std::nullopt;
+        }
+        return t + 1;
+    }
```
I'm not sure where this is going to be used, but it hints that we could use a `for_each_tablet` function.

tgrabiec replied:
The usage of this cannot be replaced with for_each_tablet(), we need iterator-like interface.

avikivity replied:
Maybe std::generator (when we move to C++23)

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+        return t + 1;
+    }
+
+    const tablet_container& tablets() const {
```
nit: noexcept

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+        return _tablets;
+    }
+
+    size_t tablet_count() const {
```
nit: noexcept

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+        return _tablets.size();
+    }
+
+    const tablet_info& get_tablet_info(token t) const {
```
nit: noexcept
(`get_tablet_id(t)` must return a valid tablet_id for any `token t`, no?)

tgrabiec replied:
Yes, we always have a mapping. But why noexcept? It's limiting to the implementation.

bhalevy replied:
otherwise, I'd expect error handling by the callers of this function.
If it doesn't throw at the moment, by design, marking it as noexcept will simplify the life of the callers.
If, in the future, we'd like to throw from this function, we then need to survey all call paths to make sure they'd handle such exception properly.

tgrabiec replied:
If the caller needs noexcept in the future, it's easy to slap noexcept because it's easy to prove this is noexcept. If we did this preemptively if there is no need, and the implementation wants to throw, we would have to prove that users don't depend on noexcept, which in general is much harder (there are many of them, and it's not obvious if they rely on it). It is also a risk to mark it as noexcept, because that's an extra constraint which we can miss to fulfill and cause abort.

tgrabiec replied:
@avikivity What's your take on this?

bhalevy replied:
Here, since all tokens must be mapped to tablets I think it should be noexcept by design.
If in the future we'd need to e.g. allocate tablet_info on-demand, I think we should expose that as `get_or_create_tablet_info` and let this function assume it is already allocated and if not terminate on internal error.

avikivity replied:
I'm not a fan of strict noexcept.

It has its place in move constructors and assignment operators, and in functions that return a future. But in random functions it's easy to add noexcept on Monday and change the implementation to something that throws on Thursday. Or even on the same day in the same patch.

Nothing original about this, I've seen the argument from @tgrabiec too.

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+    return dht::compaction_group_of(_log2_tablets, t);
+}
+
+dht::token tablet_map::get_last_token(tablet_id id) const {
```
nit: noexcept

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+    return dht::last_token_of_compaction_group(_log2_tablets, id);
+}
+
+dht::token tablet_map::get_first_token(tablet_id id) const {
```
nit: noexcept

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+    return _tablets[id];
+}
+
+tablet_id tablet_map::get_tablet_id(token t) const {
```
nit: noexcept

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+    _tablets.resize(tablet_count);
+}
+
+const tablet_info& tablet_map::get_tablet_info(tablet_id id) const {
```
nit: noexcept

(and what about range checking / assert / on_internal_error?)

tgrabiec replied:
> (and what about range checking / assert / on_internal_error?)

I'll add range checking.

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -74,6 +77,13 @@ class token_metadata_impl final {
 
+    const tablet_metadata& tablets() const { return _tablets; }
+
+    void set_tablets(tablet_metadata&& tablets) {
```
noexcept?

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -399,6 +411,14 @@ void token_metadata_impl::sort_tokens() {
+    return _impl->tablets();
+}
+
+void token_metadata::set_tablets(tablet_metadata tm) {
```
noexcept?

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -399,6 +411,14 @@ void token_metadata_impl::sort_tokens() {
     _sorted_tokens = std::move(sorted);
 }
 
+const tablet_metadata& token_metadata::tablets() const {
```
nit: noexcept

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+    _tablets[id] = std::move(map);
+}
+
+future<> tablet_metadata::clear_gently() {
```
how about `utils::clear_gently(_tablets)`?

bhalevy replied:
and also, this should be noexcept.

tgrabiec replied:
Ok.

bhalevy commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+    }
+    if (t.is_maximum()) {
+        return tablet_count() - 1;
+    }
```
doesn't `dht::compaction_group_of` take care of the above cases?
if not, then it should be fixed.
I don't see why `get_tablet_id` and `compaction_group_of` should diverge.

bhalevy commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+    // using shared pointers. We should change that and use a foreign_ptr<> to
+    // hold immutable tablet_metadata which lives on shard 0 only.
+    // See storage_service::replica_to_all_cores().
+    using table_to_map = std::unordered_map<table_id, tablet_map>;
```
I kinda dislike the name `table_to_map`.
How about `map_by_table` or `map_by_table_id`?

tgrabiec replied:
Why do you dislike it? x_to_y is a common pattern in the code base.

bhalevy replied:
then it should be called `table_to_tablet_map`, not just `table_to_map`.
The latter is unclear about what `x` is mapped to.

tgrabiec replied:
Right

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -74,6 +77,13 @@ class token_metadata_impl final {
 
     void sort_tokens();
 
+    const tablet_metadata& tablets() const { return _tablets; }
```
nit: noexcept

avikivity commented on replica/database.cc:
```diff
@@ -943,16 +951,23 @@ void database::maybe_init_schema_commitlog() {
+        erm = pt_rs->make_replication_map(schema->id(), _shared_token_metadata.get());
+    } else {
+        erm = ks.get_effective_replication_map();
+    }
```
I different approach is to change all call-sites to treat replication map creation as per-table, and adjust the creation code to hand a copy of previously created replication map if the replication strategy knows it's table-independent. It may reduce the churn.

tgrabiec replied:
That would spread the logic to many places, right now it's in a single place.

avikivity replied:
ok

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -559,7 +560,9 @@ indexed_table_select_statement::do_execute_base_query(
-    query_ranges_to_vnodes_generator ranges_to_vnodes(qp.proxy().get_token_metadata_ptr(), _schema, std::move(partition_ranges));
+    auto&& table = qp.proxy().local_db().find_column_family(_schema);
+    auto erm = table.get_effective_replication_map();
+    query_ranges_to_vnodes_generator ranges_to_vnodes(erm->make_splitter(), _schema, std::move(partition_ranges));
```
There's an existing bug where erm is not captured for the duration of the page. So if a query is slower than ring_delay, and streaming is too fast (both very unlikely) then we can query bad data.

tgrabiec replied:
I anticipated that and token_range_splitter survives erm by keeping token_metadata_ptr alive.

avikivity replied:
yes it will keep erm, but holding erm right now doesn't prevent the real topology from moving. So we have to trust ring_delay for now.

avikivity replied:
Ah, I said "erm is not captured". Well capturing it prepares the way for the full solution.

avikivity commented on dht/token.cc:
```diff
@@ -288,4 +288,14 @@ compaction_group_of(unsigned most_significant_bits, const token& t) {
     __builtin_unreachable();
 }
 
+token last_token_of_compaction_group(unsigned most_significant_bits, size_t group) {
```
Are compaction groups groups really the target here?

Tablets should work without compaction groups, at least logically.

tgrabiec replied:
Note that this is in `dht/token.cc`. This function is really about token space splitting, not really about compaction groups. Maybe we should rename those functions.

tgrabiec replied:
I added this to mirror `compaction_group_of()`, which is defined right above it. That's what I was referring to. I also use it, and they are tightly related.

avikivity replied:
Ok. It's an historical accident (compaction groups being introduced sooner). Right now it's possible for tablets and compaction groups to be split differently, but we can tighten it later and the distinction will disappear.

avikivity commented on locator/tablets.hh:
```diff
@@ -0,0 +1,177 @@
+
+std::ostream& operator<<(std::ostream&, const tablet_replica&);
+
+using tablet_replica_set = utils::small_vector<tablet_replica, 3>;
```
I'd like this to be std::span pointing into a larger space, to avoid lots of small allocations. This minor optimization isn't needed now, and the code should be well isolated.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1015,6 +1018,28 @@ schema_ptr system_keyspace::broadcast_kv_store() {
+            .with_version(generate_schema_version(id))
+            .build();
+    }();
+    return schema;
```
This has to be documented.

avikivity replied:
^^

tgrabiec replied:
It's now documented in system_keyspace.md: https://github.com/scylladb/scylladb/pull/13387/commits/ae3da41246af2d1403f42fc54f9be692ce5c155f#diff-f39affc830cb4c115f56d0101bec3e1fba14c5843cce9b4dd6a86e7f7cf46d71

avikivity commented on db/schema_tables.cc:
```diff
@@ -1275,6 +1280,13 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, std:
+        co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) -> future<> {
+            co_await db.get_notifier().update_tablet_metadata();
+        });
+    }
```
What if we have a failure later? Doesn't this have to be transactional?

tgrabiec replied:
It is transactional by the means of the group0 raft state machine. This is part of raft command (or snapshot) application, and raft will not let another read_barrier() to succeed locally before this succeeds.  

Currently, if this fails the raft implementation stops and you need a restart to trigger log replay.

avikivity replied:
This sounds bad. A bad_alloc can kill the node? We need to add retries and an automatic restart.

In any case, I believe it's in the scope of raft, not this.

avikivity commented on service/migration_manager.cc:
```diff
@@ -132,6 +133,9 @@ void migration_manager::init_messaging_service()
             cm.emplace_back(co_await db::system_keyspace::get_group0_history(proxy));
+            for (auto&& m : co_await replica::read_tablet_mutations(proxy)) {
+                cm.emplace_back(std::move(m));
+            }
```
Is this viable? A database with 100TB storage and 10GB tablets will have 10k tablets.

I guess it's okay for now, but we'll have to replace it with a stream later on.

tgrabiec replied:
Yes, we need to switch to streaming at some point. 

perf_tablets shows that 10k is fine, but with 100k we hit the commitlog mutation size limit.

avikivity replied:
ok

avikivity replied:
btw we'll hit it sooner, e.g. 20TB per datacenter * 5 datacenters, so not a huge database.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1015,6 +1018,28 @@ schema_ptr system_keyspace::broadcast_kv_store() {
+    }();
+    return schema;
+}
+
```
Is this not in all_tables()?  I don't think it's good to create it on-demand.

tgrabiec replied:
It follows the same pattern as other tables, the schema is kept in thread_local so it's initialied once.

tgrabiec commented on alternator/ttl.cc:
```diff
@@ -386,7 +386,7 @@ class token_ranges_owned_by_this_shard {
         const dht::token_range_vector _token_ranges;
      public:
-        ranges_holder_primary(const locator::effective_replication_map_ptr& erm, gms::gossiper& g, gms::inet_address ep)
+        ranges_holder_primary(const locator::token_effective_replication_map_ptr& erm, gms::gossiper& g, gms::inet_address ep)
```
This specific user cannot work with the abstract class, it is token-specific.

avikivity replied:
Ok. Maybe we should call it vnode_effective_replication_map_ptr? "token" is a building block that is used by both vnodes and tablets.

tgrabiec commented on locator/abstract_replication_strategy.cc:
```diff
@@ -368,13 +368,21 @@ effective_replication_map::~effective_replication_map() {
 }
 
-effective_replication_map::factory_key effective_replication_map::make_factory_key(const abstract_replication_strategy::ptr_type& rs, const token_metadata_ptr& tmptr) {
+effective_replication_map::effective_replication_map(abstract_replication_strategy::ptr_type rs,
```
ACK.

bhalevy commented on service/storage_service.cc:
```diff
@@ -2244,10 +2246,27 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
             auto& db = ss._db.local();
             for (auto& ks_name : keyspaces) {
                 auto rs = db.find_keyspace(ks_name).get_replication_strategy_ptr();
+                if (rs->is_per_table()) {
```
Yes. A preliminary loop clones shard 0's token_metadata on all other shards
and then it proceeds to the next step which is generating the modified e_r_m using `get_erm_factory().create_effective_replication_map`

Shard 0 gets the result from this call
and the other shards are expected to find it in the erm factory cache.

bhalevy replied:
Shouldn't the tablets map be part of the e_r_m?

tgrabiec replied:
> i.e. the per-table erms could be stored in the per-keyspace erm.

It seems cleaner to store them directly in the `table` object because that's where they really belong (they're per table). We avoid double lookup this way, you only lookup the table and get the erm from there, instead of first the keyspace and then the table in that keyspace.

Extending the per-keyspace erm with per-table lookup capabilities seemed like unnecessary complication in the end.

tgrabiec replied:
> Shard 0 gets the result from this call
> and the other shards are expected to find it in the erm factory cache.

If we cloned directly from here, would it mean we don't need the cache?

tgrabiec replied:
> Keeping the tablets map separately means we'll need to come up with an equivalent mechanism just for tablets, while we can reuse the existing mechanism easily.

Tablet maps are kept in token_metadata, not separately. Do you mean per-table erms? What kind of tracking are we talking about?

tgrabiec replied:
> > If we cloned directly from here, would it mean we don't need the cache?
> 
> I think we clone in the cache also when creating the initial version from

Right.

tgrabiec replied:
> IIUC, the tablet maps contain the full replication information. Does it even have a notion of a primary owner of a tablet?

It doesn't.

> If so, I think they should live as a member in the e_r_m, rather in the token_metadata object.

erm = rs_instance + token_metadata

We cannot make erm the primary owner of tablet_map, because erm is computed, based on other structures with longer lifetime. We would have to manage tablet_metadata like we manage token_metadata, and make replication_strategy interfaces accept both. It's just simpler to make tablet_metadata a member of token_metadata, and say that token_metadata is all metadata related to topology which is not computed based on options, but fundamental (read from system tables). It already is, it stores rack and dc assignment, not just tokens. So it can also store tablet metadata.

tgrabiec replied:
> > IIUC, the tablet maps contain the full replication information. Does it even have a notion of a primary owner of a tablet?

> It doesn't.

I think I misunderstood the question. The owner of tablet_map is token_metadata (it's an embedded member, not a pointer). tablet_map is versioned together with token_metadata. New version of table_map is created by creating a new version of token_metadata.

avikivity replied:
So it's looked up in token_metadata or one of the similar-lifetime objects, not table?

tgrabiec replied:
Yes, erm looks up in tablet_metadata. This could be optimized by storing a pointer in the erm.

tgrabiec commented on replica/tablets.cc:
```diff
@@ -0,0 +1,181 @@
+        ++tid;
+        co_await coroutine::maybe_yield();
+    }
+    co_return std::move(m);
```
There's no copy elision for coroutines, so it's needed to avoid a copy. It's a common pattern in our code base.

tgrabiec commented on replica/tablets.cc:
```diff
@@ -0,0 +1,181 @@
+
+future<> save_tablet_metadata(replica::database& db, const tablet_metadata& tm, api::timestamp_type ts) {
+    tablet_logger.trace("Saving tablet metadata: {}", tm);
+    std::vector<mutation> muts;
```
Ok.

tgrabiec commented on replica/tablets.cc:
```diff
@@ -0,0 +1,181 @@
+            auto tablet_count = row.get_as<int>("tablet_count");
+            current = active_tablet_map{table, tablet_map(tablet_count), 0};
+            if (current->map.tablet_count() != tablet_count) {
+                throw std::runtime_error(format("tablet count mismatch between on-disk and in-memory metadata for table {} tablet {}: {} != {}",
```
tablet_map constructor can round up the count.

tgrabiec replied:
Yeah, it's probably better to require callers to provide the correct count. I'll change it to verify rather than round up.

tgrabiec replied:
In the constructor of tablet_map, that is.

tgrabiec commented on replica/tablets.hh:
```diff
@@ -0,0 +1,63 @@
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
```
I can move schema construction to tablets.cc, then all the details will be contained in tablets.cc, system_keyspace will only know about tablets by selecting the table in `all_tables()`.

tgrabiec commented on main.cc:
```diff
@@ -1148,6 +1148,11 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             qp.start(std::ref(proxy), std::ref(forward_service), std::move(local_data_dict), std::ref(mm_notifier), std::ref(mm), qp_mcfg, std::ref(cql_config), std::move(auth_prep_cache_config), std::ref(group0_client)).get();
+
+            ss.invoke_on_all([&] (service::storage_service& ss) {
+                ss.set_query_processor(qp.local());
```
ss.start() is called before qp.start(), so qp is not ready when we invoke the storage_service constructor. From the point of view of the implementation of storage_service, taking something by the constructor means that this thing should be ready to use. It's not just about passing the pointer. 

We used to have those pointers globally available to everyone and it created a mess because you couldn't know at usage site if is actually ready to use. 

The set_query_processor() thing is ugly, but at least you know in the implementation that you cannot trust it and be extra careful.

gusev-p replied:
> From the point of view of the implementation of storage_service, taking something by the constructor means that this thing should be ready to use. It's not just about passing the pointer.

Agree, I just missed that `storage_service` was initialized much earlier.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -179,4 +181,129 @@ size_t tablet_metadata::external_memory_usage() const {
+            if (!fs.tablets) {
+                throw exceptions::configuration_exception("Tablet replication is not enabled");
+            }
+            parse_initial_tablets(c.second);
```
It's there so that abstract_replication_strategy::validate_options() validates all the options. The question is more general actually, why do we need validate_options() if we parse the options in the constructor anyway, and it applies not only to tablet options, but other options as well. I don't have an answer for that, I just played along.

tgrabiec commented on locator/abstract_replication_strategy.hh:
```diff
@@ -126,10 +147,64 @@ public:
+/// It will be accessed only if the replication strategy actually works in per-table mode,
+/// that is after mark_as_per_table() is called, and as a result
+/// abstract_replication_strategy::is_per_table() returns true.
+class per_table_replication_strategy {
```
It's to avoid the "diamond problem" in case replication strategy implements multiple traits.

tgrabiec replied:
> And why not make `is_per_table()` and `maybe_as_per_table()` virtual functions, which would be overriden in `per_table_replication_strategy`? (then we could remove `_per_table` bool) Is it due to performance considerations?

This is to avoid the cost of virtual calls.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -0,0 +1,309 @@
+}
+
+std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet_options() const {
+    std::unordered_set<sstring> opts;
```
Cutting options from _config_options is not safe in general because they are exposed publicly, and accessed after construction.

tgrabiec replied:
You mean avoiding calling recognized_tablet_options() from the constructor of network_topology_strategy::network_topology_strategy? Yes, I think we could do that.

gusev-p replied:
It's strange to me that parsing one simple option `initial_tablets` takes one and a half pages of code and three mentions of the word `"initial_tablets"`. It should be something like
```
void parse_options(strategy_option& options) {
    if (options.try_get("initial_tablets", _initial_tablets)) {
        if (!fs.tablets) {
            throw exceptions::configuration_exception("Tablet replication is not enabled");
        }
    }
}
```
but I guess it's out of scope of this PR

tgrabiec commented on test/topology/test_tablets.py:
```diff
@@ -0,0 +1,94 @@
+
+    keys = range(100)
+    for k in keys:
+        manager.cql.execute(f"INSERT INTO test.test (pk, c) VALUES ({k}, {k});")
```
Thanks, will use.

kbr-scylla commented on locator/network_topology_strategy.cc:
```diff
@@ -268,7 +281,52 @@ void network_topology_strategy::validate_options(const gms::feature_service&) co
+                token_range = tm->ring_range(dht::minimum_token());
+            }
+            if (tracker.add_endpoint_and_check_if_done(ep)) {
+                break;
```
Won't this loop forever if there are not enough nodes in one of the DCs?

tgrabiec replied:
No, because `rf` for a dc is reduced in the tracker if there is not enough nodes (`std::min(rf, node_count)`):

```
data_center_endpoints(size_t rf, size_t rack_count, size_t node_count, endpoint_set& endpoints, endpoint_dc_rack_set& racks)
            : _endpoints(endpoints)
            , _racks(racks)
            // If there aren't enough nodes in this DC to fill the RF, the number of nodes is the effective RF.
            , _rf_left(std::min(rf, node_count))
      
```

tgrabiec commented on service/migration_manager.cc:
```diff
@@ -738,14 +738,17 @@ future<std::vector<mutation>> migration_manager::prepare_aggregate_drop_announce
     auto& keyspace = db.find_keyspace(ks_name);
     mlogger.info("Drop Keyspace '{}'", ks_name);
-    return db::schema_tables::make_drop_keyspace_mutations(db.features().cluster_schema_features(), keyspace.metadata(), ts);
+    return seastar::async([this, &db, &keyspace, ts, ks_name] {
```
Yes, sloppy commit splitting. I'll move it.

tgrabiec commented on cql3/statements/select_statement.cc:
```diff
@@ -45,6 +45,7 @@
 #include "utils/result.hh"
```
Yes, it should be "around tokens in the ring" to be precise, or vnodes.

tgrabiec replied:
Although, "ring" could be also confusing because what tablets partition is also a ring, we just have many rings now. So perhaps we should stick to vnodes, or "tokens in the global ring"? Other suggestions?

kbr-scylla replied:
I'm for "vnodes" (and "vnode boundaries") vs "tablets" (and "tablet boundaries")

tgrabiec commented on main.cc:
```diff
@@ -1208,6 +1208,10 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
                     throw bad_configuration_error();
                 }
+                if (cfg->check_experimental(db::experimental_features_t::feature::TABLETS)) {
+                    startlog.error("Bad configuration: RAFT feature has to be enabled if TABLETS is enabled");
```
Currently they don't depend on the topology coordinator, they will in the future when they're integrated with topology changes.

I'll update the message.

tgrabiec commented on locator/abstract_replication_strategy.hh:
```diff
@@ -128,11 +128,48 @@ public:
+    const topology& get_topology() const noexcept { return _tmptr->get_topology(); }
+    size_t get_replication_factor() const { return _replication_factor; }
+
+    virtual inet_address_vector_replica_set get_natural_endpoints(const token& search_token) const = 0;
```
Ok

tgrabiec commented on locator/tablets.cc:
```diff
@@ -0,0 +1,125 @@
+}
+
+void tablet_metadata::set_tablet_map(table_id id, tablet_map map) {
+    _tablets[id] = std::move(map);
```
Yes.

tgrabiec replied:
Actually no, because _tablets is a chunked_vector.

tgrabiec commented on replica/tablets.cc:
```diff
@@ -0,0 +1,181 @@
+        if (auto tr_info = tablets.get_tablet_transition_info(tid)) {
+            m.set_clustered_cell(ck, "new_replicas", make_set_value(replica_set_type, replicas_to_data_value(tr_info->next)), ts);
+        } else {
+            m.set_clustered_cell(ck, *s->get_column_definition("new_replicas"), atomic_cell::make_dead(ts, gc_now));
```
The overload which takes column by name creates a live cell, in the else branch we create a dead cell.

tgrabiec replied:
Oh, but we don't have to, because there's partition tombstone.

tgrabiec commented on replica/tablets.cc:
```diff
@@ -0,0 +1,181 @@
+            current = active_tablet_map{table, tablet_map(tablet_count), 0};
+            if (current->map.tablet_count() != tablet_count) {
+                throw std::runtime_error(format("tablet count mismatch between on-disk and in-memory metadata for table {} tablet {}: {} != {}",
+                                                table, current->map.tablet_count(), current->tid, tablet_count));
```
right.

tgrabiec commented on db/config.hh:
```diff
@@ -91,6 +91,7 @@ struct experimental_features_t {
         RAFT,
         BROADCAST_TABLES,
         KEYSPACE_STORAGE_OPTIONS,
+        TABLETS,
```
Right.

tgrabiec commented on tombstone_gc.cc:
```diff
@@ -166,9 +166,8 @@ static bool needs_repair_before_gc(const replica::database& db, sstring ks_name)
-    auto erm = ks.get_effective_replication_map();
     bool needs_repair = rs.get_type() != locator::replication_strategy_type::local
-            && erm->get_replication_factor() != 1;
+            && rs.get_replication_factor(db.get_token_metadata()) != 1;
```
For keyspaces which use tablets there is no per-keyspace erm, so we can't get the replication factor from there. We can get it from rs, which works for all keyspaces.

tgrabiec replied:
We can't use the per-table erm here because this is a validation which is executed before the table is created.

tgrabiec commented on alternator/ttl.cc:
As to why "token_", what I had in mind was that the replication is derived from the token ring. All replication strategies use tokens at the interface, but only the old one is derived from the old token metadata.  

You could also say that tablets use the ring, because the domain for tokens is the ring, so tablets also use the ring.

tgrabiec replied:
I'm ok with "vnode", can we settle on that?

tgrabiec commented on main.cc:
```diff
@@ -1148,6 +1148,11 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             qp.start(std::ref(proxy), std::ref(forward_service), std::move(local_data_dict), std::ref(mm_notifier), std::ref(mm), qp_mcfg, std::ref(cql_config), std::move(auth_prep_cache_config), std::ref(group0_client)).get();
+
+            ss.invoke_on_all([&] (service::storage_service& ss) {
+                ss.set_query_processor(qp.local());
```
In general yes, but qp is not stopped currently, so there's no need to.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/547/

tgrabiec commented on locator/tablets.hh:
```diff
@@ -0,0 +1,231 @@
+    /// Constructs a tablet map.
+    ///
+    /// \param tablet_count The desired tablets to allocate. Must be a power of two.
+    explicit tablet_map(size_t tablet_count);
```
For some users it would, for some it would be the opposite (we would push the check and transformation to them). The persistence layer reads the count from the system table. I don't want to store log2 size there because that's an ABI and it would limit us in the future when we relax this constraint.

tgrabiec commented:
In v2:

- Changed the terminology from token-based replication strategy to vnode-based replication strategy 

- Moved power-of-2 check to the tablet_map constructor

- Updated the feature-check message to mention consistent_cluster_management instead of RAFT.

- Documented get_natural_endpoints() et al.

- Use insert_or_assign() instead of operator[]

- Fixed typo in replicate_to_all_cores()

- Drop unnecessary deletion of new_replicas

- Preserved noexcept on effective_replication_map interface

- Simplify option processing in the constructor of network_topology_staretgy by making table_aware_replication_strategy options it recognizes

- Added a type hint to driver_connect() in pylib

- Made tablet_id an enum class for safety

- Introduced tablet_map::first_tablet() / tablet_map::last_tablet()

- Added validation that pending.size() == 1 when reading tablet metadata

- Added range checks for tablet id in tablet_map accessors

- Added more documentation to tablet_map

- Added a test case to NetworkTopologyStrategy_tablets_test for the case of not enough nodes in a DC to meet RF.

- Start from a random token when allocating tablets to spread tablets more evenly

- Warn when rounding up tablet count in allocate_tablets_for_new_table()

- Use last_token_of_compaction_group() in split_token_range_msb()

- Improved documentation of token_range_splitter

- [patch splitting] Moved introduction of std::async() in prepare_keyspace_drop_announcement() to the right patch

Not addressed:

- fmt printers

- Did not add noexcept to new methods

tgrabiec commented:
In v3:
   - Integration with raft-based topology: https://github.com/scylladb/scylladb/pull/13387/commits/e35269f1c1b3ffb4ddcc674a3d4e6592613a19f6
   - Fixed failures after rebasing on top of topology nodes: https://github.com/scylladb/scylladb/pull/13498

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/607/

tgrabiec commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/607/

https://github.com/scylladb/scylladb/issues/13522

tgrabiec commented:
In V3:
  - disable experimental raft in tablets_test to avoid https://github.com/scylladb/scylladb/issues/13522
  - fix test_tablets.py to be immune to slow schema propagation which sometimes causes the test to fail
  - rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/692/

avikivity commented:
Merging gives me this:

```
[1214/2727]  boost     dev   [ FAIL ] boost.locator_topology_test.test_update_node.1
[1221/2727]  boost     dev   [ PASS ] boost.map_difference_test.differing_values.1                                                  
[1222/2727]  boost     dev   [ FAIL ] boost.locator_topology_test.test_add_node.1
```

So I probably botched the merge. Please rebase.

kbr-scylla commented:
> Merging

But I'm in the middle of (re)review! :(

avikivity commented:
> > Merging
> 
> But I'm in the middle of (re)review! :(

Holding (in any case we need a rebase).

kbr-scylla reviewed (approved):
Just minor questions/comments

note that there's a new version of #13438

also, I didn't review the prefix of this PR which is #13498 - we should probably wait until the review of that one finishes before we merge this one.

tgrabiec commented on replica/database.cc:
```diff
@@ -1151,6 +1151,18 @@ std::vector<sstring> database::get_non_local_strategy_keyspaces() const {
+    res.reserve(_keyspaces.size());
+    for (auto const& i : _keyspaces) {
+        auto&& rs = i.second.get_replication_strategy();
+        if (rs.get_type() != locator::replication_strategy_type::local && !rs.is_per_table()) {
```
I can add is_vnode_based() to abstract_replication_strategy to embed this assumption in the rs layer.

tgrabiec commented on replica/database.cc:
```diff
@@ -1167,7 +1167,8 @@ std::unordered_map<sstring, locator::vnode_effective_replication_map_ptr> databa
     std::unordered_map<sstring, locator::vnode_effective_replication_map_ptr> res;
```
I thought name change is not necessary, since the you are asking for per-keyspace erms, so you only get erms which are per-keyspace (not per table).

gleb-cloudius replied:
I do not see where is filters out per table keyspaces.

gleb-cloudius replied:
> I do not see where is filters out per table keyspaces.

Ah, it is in the next patch. Should they be squashed?

tgrabiec replied:
Which patches?

gleb-cloudius replied:
It looks to me like `db: Introduce get_non_local_vnode_based_strategy_keyspaces() ` and  `db: Exclude keyspace with per-table replication in get_non_local_strategy_keyspaces_erms()` have very similar purpose. But this is just a nitpick.

tgrabiec replied:
Similar purpose, yes, but a bit different solution (one introduces a new method, the other changes behavior of an existing one). Can be enough justification for a split.

tgrabiec commented on dht/token.hh:
```diff
@@ -218,6 +218,16 @@ inline bool operator<=(const token& t1, const token& t2) { return std::rel_ops::
+// Precondition: t.kind() == dht::token::kind::key
+inline
+token next_token(const token& t) {
+    return {dht::token::kind::key, t._data + 1};
```
This is specified as a precondition:

```
The caller must ensure there is a next token, otherwise
the result is unspecified.
```

If _data is max int, there is no next token, so precondition is not satisfied.

tgrabiec commented on dht/i_partitioner.cc:
```diff
@@ -457,28 +457,22 @@ dht::token_range_vector split_token_range_msb(unsigned most_significant_bits) {
-                        i, most_significant_bits, token, compaction_group_of(most_significant_bits, token)));
+                                                       i, most_significant_bits, token, compaction_group_of(most_significant_bits, token)));
             }
-            end_bound = dht::token_range::bound(token, false);
```
I think users of this function don't assume that the range is in the form (x, y].

tgrabiec commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -119,10 +131,10 @@ void endpoints_check(
-        BOOST_CHECK(rf.second == nts_ptr->get_replication_factor(rf.first));
+    for (auto&& [dc, rf] : dc_rf) {
+        auto i = nodes_per_dc.find(dc);
+        auto effective_rf = i == nodes_per_dc.end() ? 0 : std::min<size_t>(nts_ptr->get_replication_factor(dc), i->second.size());
```
> Is the i == nodes_per_dc.end() check needed?

I think you're right, it's not needed. Will check.

> Maybe it was never used in a test where a DC had fewer nodes than RF?

It wasn't.

tgrabiec commented on locator/tablets.hh:
```diff
@@ -0,0 +1,231 @@
+    std::unordered_map<tablet_id, tablet_transition_info> _transitions;
+public:
+    /// Default-constructed tablet_map is invalid and should not be accessed to obtain information.
+    /// Provided for so that it can be used in std::unordered_map<>.
```
Right, will change.

tgrabiec commented on replica/database.hh:
```diff
@@ -758,17 +761,19 @@ public:
-        : table(schema, std::move(cfg), std::move(sopts), &cl, cm, sm, cl_stats, row_cache_tracker) {}
-    table(schema_ptr schema, config cfg, lw_shared_ptr<const storage_options> sopts, no_commitlog, compaction_manager& cm, sstables::sstables_manager& sm, cell_locker_stats& cl_stats, cache_tracker& row_cache_tracker)
-        : table(schema, std::move(cfg), std::move(sopts), nullptr, cm, sm, cl_stats, row_cache_tracker) {}
+    table(schema_ptr schema, config cfg, lw_shared_ptr<const storage_options> sopts, db::commitlog& cl, compaction_manager& cm, sstables::sstables_manager& sm, cell_locker_stats& cl_stats, cache_tracker& row_cache_tracker, locator::effective_replication_map_ptr erm = nullptr)
```
There is a bunch of test code which creates table objects in isolation for testing, which cannot provide a valid erm. There is no replication there, just like there is no holding keyspace. All production code does provide it, so callers don't have to check for null. 

The table object really has two parts, the replica side and coordinator side. They're mixed, but some tests only use the replica side of it, without bothering with replication.

tgrabiec replied:
I could drop the default and pass nullptr in tests.

tgrabiec commented:
In V4:
  - rebased
  - Introduce abstract_replicaiton_strategy::is_vnode_based() and use in get_non_local_vnode_based_strategy_keyspaces()
  - Drop unneeded check for existence of dc in network_topology_strategy_test
  - Don't use unordered_map::operator[] on _tablets which allows me to drop the default tablet_map constructor
  - Don't default erm to nullptr in table constructor, pass nullptr in tests

gleb-cloudius commented on locator/abstract_replication_strategy.cc:
```diff
@@ -368,6 +368,14 @@ vnode_effective_replication_map::~vnode_effective_replication_map() {
     }
 }
 
+effective_replication_map::effective_replication_map(abstract_replication_strategy::ptr_type rs,
```
In the commit message `s/token_effective_replication_map/vnode_effective_replication_map/`

tgrabiec replied:
Fixed.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/823/

gleb-cloudius commented on service/migration_manager.cc:
```diff
@@ -132,6 +133,9 @@ void migration_manager::init_messaging_service()
             cm.emplace_back(co_await db::system_keyspace::get_group0_history(proxy));
+            for (auto&& m : co_await replica::read_tablet_mutations(proxy)) {
+                cm.emplace_back(std::move(m));
+            }
```
Why move it as part of schema and not topology mutations?

tgrabiec replied:
When creating a table or dropping a table, I want tablet mutations and schema mutations to be applied atomically as part of one command. This way schema changes can be a single command.

tgrabiec replied:
In other cases, like topology changes, tablet mutations can be part of topology mutations.

tgrabiec replied:
Actually, we should prohibit pulling schema outside of raft if we care about local state machine consistency. It's not enough to pull tablets, we also need topology, on which those tablets depend, to properly choose replicas. For example, which ones are dc local.

If we're talking tablet mutations in the migration request, I can actually move them to the topology pull. It's just that it wasn't yet available at the time this was implemented.

gleb-cloudius replied:
> Actually, we should prohibit pulling schema outside of raft if we care about local state machine consistency. It's not enough to pull tablets, we also need topology, on which those tablets depend, to properly choose replicas. For example, which ones are dc local.

Yes. We should at least do `raft_read_barrier`, recheck if schema is available and then pull. I think we still need to pull since the schema we want to pull may be the old one and not the latest one, no?

tgrabiec replied:
But  should change the order, and sync raft first.

gleb-cloudius replied:
What is the former? The one in maybe_sync? So what should it do?

```
co_await read_barrier()
return get_schema_definition()
```
?

tgrabiec replied:
> What is the former? The one in maybe_sync? So what should it do?
> 
> ```
> co_await read_barrier()
> return get_schema_definition()
> ```
> 
> ?

Yes

tgrabiec replied:
But we should also avoid read_barrier() if we already know the schema version.

gleb-cloudius replied:
So what I wrote here https://github.com/scylladb/scylladb/pull/13387#discussion_r1175130799:

```
if (schema is not available) {
  co_await read_barrier()
}

if (schema is not available) {
  return get_schema_definition()
}
```

tgrabiec replied:
> So what I wrote here [#13387 (comment)](https://github.com/scylladb/scylladb/pull/13387#discussion_r1175130799):
> 

get_schema_definition() does the check, so:

```
 if (auto s = local_schema_registry().get_or_null(v)) {
     return s; 
 }
 co_await read_barrier();
 return get_schema_definition(v, ...);
```

tgrabiec replied:
And maybe_sync() becomes mark_synced(), since we did sync via raft.

tgrabiec replied:
We actually can drop maybe_sync() altogether, since everything will be synced before being registered.

tgrabiec replied:
> Why move it as part of schema and not topology mutations?

Pulling it as part of topology mutations is more tricky because we need to apply and reflect topology before the schema. So I propose we do it incrementally.

tgrabiec commented:
@gleb-cloudius Are there any merge blockers?

gleb-cloudius commented:
> @gleb-cloudius Are there any merge blockers?

Not from my side.

tgrabiec commented:
@kbr-scylla Please review / merge.

kbr-scylla commented:
Queued
there was a minor merge conflict but easy to solve

avikivity commented:
Hurray!
